### Python - Built-in Modules

The Python interactive shell has a number of built-in functions. They are loaded automatically as a shell starts and are always available, such as `print()` and `input()` for I/O, number conversion functions `int()`, `float()`, `complex()`, data type conversions `list()`, `tuple()`, `set()`, etc.

In addition to built-in functions, a large number of pre-defined functions are also available as a part of libraries bundled with Python distributions. These functions are defined in modules are called built-in modules.

Built-in modules are written in `C` and integrated with the `Python shell`. Each `built-in module `contains resources for certain system-specific functionalities such as `OS` management, disk IO, etc. The standard library also contains many Python scripts (with the `.py` extension) containing useful utilities.

To display a list of all available modules, use the following command in the Python console:

In [2]:
help("modules")


Please wait a moment while I gather a list of all available modules...

IPython             bleach              mimetypes           stack_data
Mypackage           bs4                 mistune             stat
PIL                 builtins            mmap                statistics
__future__          bz2                 mmapfile            string
_abc                cProfile            mmsystem            stringprep
_aix_support        calc                modulefinder        struct
_argon2_cffi_bindings calc2               msilib              subprocess
_ast                calendar            msvcrt              sunau
_asyncio            certifi             multiprocessing     symbol
_bisect             cffi                nbclient            symtable
_blake2             cgi                 nbconvert           sys
_bootlocale         cgitb               nbformat            sysconfig
_bootsubprocess     chunk               nest_asyncio        tabnanny
_bz2                cmath            

### 1. Python - OS Module


It is possible to automatically perform many operating system tasks. The `OS` module in Python provides functions for creating and removing a directory (folder), fetching its contents, changing and identifying the current directory, etc.

You first need to import the os module to interact with the underlying operating system. So, `import` it using the `import os` statement before using its functions.

#### Getting Current Working Directory
The `getcwd()` function confirms returns the current working directory.

In [3]:
import os

In [4]:
os.getcwd()

'E:\\Python_world\\python_amulya-s_academy_practice\\magic_methods_and_all'

### Creating a Directory
We can create a new directory using the `os.mkdir()` function, as shown below.


In [5]:
import os

os.mkdir("MyPythonProject")

`A new directory corresponding to the path in the string argument of the function will be created. If you open the C:\ drive, then you will see the MyPythonProject folder has been created.`

By default, if you don't specify the whole path in the `mkdir()` function, it will create the specified directory in the current working directory or drive. The following will create MyPythonProject in the C:\Python37 directory.

In [1]:
import os
os.getcwd()

'E:\\Python_world\\python_amulya-s_academy_practice\\magic_methods_and_all'

### Changing the Current Working Directory
We must first change the current working directory to a newly created one before doing any operations in it. This is done using the `chdir()` function. The following change current working directory to `'E:\\Python_world\\python_amulya-s_academy_practice\\magic_methods_and_all'\\MyPythonProject`.

In [2]:
import os
# changing the current directory
os.chdir("E:\\Python_world\\python_amulya-s_academy_practice\\magic_methods_and_all\\MyPythonProject")

In [3]:
os.getcwd()

'E:\\Python_world\\python_amulya-s_academy_practice\\magic_methods_and_all\\MyPythonProject'

`You can change the current working directory to a drive. The following makes the E:\ drive as the current working directory.`



In [4]:
os.chdir("E:\\")

In [5]:
os.getcwd()

'E:\\'

#### Note :

In order to set the current directory to the parent directory use `".."` as the argument in the `chdir()` function.



In [6]:
import os
os.chdir("..")

In [7]:
os.getcwd()

'E:\\'

In [8]:
os.chdir('E:\\Python_world\\python_amulya-s_academy_practice\\magic_methods_and_all')

In [9]:
os.getcwd()

'E:\\Python_world\\python_amulya-s_academy_practice\\magic_methods_and_all'

In [10]:
os.chdir("..")

In [11]:
os.getcwd()

'E:\\Python_world\\python_amulya-s_academy_practice'

In [12]:
import os

os.chdir(".\\") # to back one dicrectory back 

In [13]:
os.getcwd()

'E:\\Python_world\\python_amulya-s_academy_practice'

In [14]:

os.chdir('E:\\Python_world\\python_amulya-s_academy_practice\\magic_methods_and_all')

In [15]:
import os
os.getcwd()

'E:\\Python_world\\python_amulya-s_academy_practice\\magic_methods_and_all'

In [17]:
os.chdir("..\\") # to back one dicrectory back 

In [18]:
os.getcwd()

'E:\\Python_world\\python_amulya-s_academy_practice'

In [19]:
os.chdir("./magic_methods_and_all") # for forward one directory

In [20]:
os.getcwd()

'E:\\Python_world\\python_amulya-s_academy_practice\\magic_methods_and_all'

In [ ]:
# os.chmod use  for read write execute 

### 2. Python - sys Module

The sys module provides functions and variables used to manipulate different parts of the Python runtime environment. You will learn some of the important features of this module here.

#### sys.argv
`sys.argv` returns a list of command line arguments passed to a Python script. The item at index 0 in this list is always the name of the script. The rest of the arguments are stored at the subsequent indices.

Here is a Python script (test.py) consuming two arguments from the command line.

This script is executed from command line as follows

`E:\Python_world\python_amulya-s_academy_practice\magic_methods_and_all>python test_sys.py Python c# Java`
>>`You entered:  Python c# Java`

Above, `sys.argv[1]` contains the first argument 'Python',` sys.argv[2] `contains the second argument 'Python', and `sys.argv[3]` contains the third argument 'Java'. `sys.argv[0]` contains the script file name `test.py`.

#### sys.exit
This causes the script to exit back to either the Python console or the command prompt. This is generally used to safely exit from the program in case of generation of an exception.

#### sys.maxsize
Returns the largest integer a variable can take.

In [23]:
import sys

In [24]:
sys.maxsize

9223372036854775807

In [26]:
len(str(sys.maxsize))

19

#### sys.path
This is an environment variable that is a search path for all Python modules.

In [27]:
import sys

sys.path

['E:\\Python_world\\python_amulya-s_academy_practice\\magic_methods_and_all',
 'C:\\Users\\ankit.shukla\\AppData\\Local\\Programs\\Python\\Python39\\python39.zip',
 'C:\\Users\\ankit.shukla\\AppData\\Local\\Programs\\Python\\Python39\\DLLs',
 'C:\\Users\\ankit.shukla\\AppData\\Local\\Programs\\Python\\Python39\\lib',
 'C:\\Users\\ankit.shukla\\AppData\\Local\\Programs\\Python\\Python39',
 '',
 'C:\\Users\\ankit.shukla\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packages',
 'C:\\Users\\ankit.shukla\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packages\\win32',
 'C:\\Users\\ankit.shukla\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packages\\win32\\lib',
 'C:\\Users\\ankit.shukla\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packages\\Pythonwin']

#### sys.version
This attribute displays a string containing the version number of the current Python interpreter.

In [29]:
import sys
sys.version

'3.9.7 (tags/v3.9.7:1016ef3, Aug 30 2021, 20:19:38) [MSC v.1929 64 bit (AMD64)]'

In [30]:
sys.version_info

sys.version_info(major=3, minor=9, micro=7, releaselevel='final', serial=0)

### 3. Python - Math Module

Some of the most popular mathematical functions are defined in the math module. These include trigonometric functions, representation functions, logarithmic functions, angle conversion functions, etc. In addition, two mathematical constants are also defined in this module.

`Pi is a well-known mathematical constant, which is defined as the ratio of the circumference to the diameter of a circle and its value is 3.141592653589793.`

In [31]:
import math 

math.pi

3.141592653589793

Another well-known mathematical constant defined in the math module is e. It is called `Euler's number` and it is a base of the `natural logarithm`. Its value is `2.718281828459045`.

`The math module contains functions for calculating various trigonometric ratios for a given angle. The functions (sin, cos, tan, etc.) need the angle in radians as an argument. We, on the other hand, are used to express the angle in degrees. The math module presents two angle conversion functions: degrees() and radians(), to convert the angle from degrees to radians and vice versa. For example, the following statements convert the angle of 30 degrees to radians and back (Note: π radians is equivalent to 180 degrees).`



In [35]:
import math 

math.radians(30)

0.5235987755982988

In [36]:
math.degrees(math.pi/6)

29.999999999999996

`The following statements show sin, cos and tan ratios for the angle of 30 degrees (0.5235987755982988 radians):`

In [37]:
import math
math.sin(0.5235987755982988)

0.49999999999999994

In [38]:
math.cos(0.5235987755982988)

0.8660254037844387

In [40]:
math.sin(0.5235987755982988)

0.49999999999999994

You may recall that sin(30)=0.5, cos(30)=32 (which is 0.8660254037844387) and tan(30)= 13 (which is 0.5773502691896257).

#### math.log()
`The math.log() method returns the natural logarithm of a given number. The natural logarithm is calculated to the base e.`



In [41]:
import math
math.log(10)

2.302585092994046

### math.log10()
`The math.log10() method returns the base-10 logarithm of the given number. It is called the standard logarithm.`

In [42]:
import math 

math.log10(10)

1.0

#### math.exp()
`The math.exp() method returns a float number after raising e to the power of the given number. In other words, exp(x) gives e**x.`



In [43]:
import math

math.exp(10)

22026.465794806718

In [44]:
math.e**10

22026.465794806703

#### math.pow()
`The math.pow() method receives two float arguments, raises the first to the second and returns the result. In other words, pow(4,4) is equivalent to 4**4.`

In [45]:
import math 

math.pow(2,4)

16.0

In [46]:
2**4

16

#### math.sqrt()
`The math.sqrt() method returns the square root of a given number.`


In [47]:
import math

math.sqrt(100)

10.0

In [48]:
math.sqrt(3)

1.7320508075688772

### 4. Python - Statistics Module

`The statistics module provides functions to mathematical statistics of numeric data. The following popular statistical functions are defined in this module.`

### Mean
The `mean()` method calculates the arithmetic mean of the numbers in a list.

In [1]:
import statistics

statistics.mean([5, 6, 9])

6.666666666666667

In [2]:
statistics.mean((20, 30 ))

25

In [3]:
type(statistics.mean((20, 30 )))

int

### Median
The `median()` method returns the middle value of numeric data in a list.

In [4]:
import statistics

statistics.median([2,3,4,5,6,2,1,2,3])

3

#### Mode
The `mode()` method returns the most common data point in the list.

In [5]:
import statistics
statistics.mode([2,5,3,2,8,3,9,4,2,5,6])


2

In [6]:
statistics.mode((2,3,3,4,5,6,5,6,4,2,2,))

2

In [7]:
statistics.mode((2,3,3,2,3,2))

2

#### Standard Deviation
The `stdev()` method calculates the standard deviation on a given sample in the form of a list.

In [8]:
import statistics
statistics.stdev([1,1.5,2,2.5,3,3.5,4,4.5,5])

1.3693063937629153

In [9]:
statistics.stdev((12,23,23,11,23,34))

8.508818954473059

In [10]:
statistics.stdev((12,10,10,12,10.5,11.5))

0.9486832980505138

### 5. Python - Collections Module

The collections module provides alternatives to built-in container data types such as list, tuple and dict.


###  namedtuple()

The `namedtuple()` function returns a tuple-like object with named fields. These field attributes are accessible by lookup as well as by index.

General usage of this function is:

`syntax: collections.namedtuple(type_name, field-list)`

In [13]:
import collections

student = collections.namedtuple('student', ['name', 'age', 'marks'])

To create a new object of this namedtuple, do the following:

In [14]:
# Create New Object of Namedtuple

s1 = student("Imran", 21, 98)

In [15]:
s1.name

'Imran'

In [16]:
s1[0]

'Imran'

In [17]:
s1.age

21

In [18]:
s1.marks

98

In [19]:
s1[2]

98

In [20]:
s1[1]

21

In [21]:
s1[4]

IndexError: tuple index out of range

In [72]:
import collections

student12 = collections.namedtuple('student', ['name', 'age', 'marks'])

In [73]:
s1 = student12("Imran", 21, 98)

In [74]:
s1.name

'Imran'

In [75]:
s1.age

21

In [76]:
s1.marks

98

`Point = namedtuple('Point', ['x', 'y'])`

`Point.__doc__                   # docstring for the new class`

`'Point(x, y)'`

`p = Point(11, y=22)             # instantiate with positional args or keywords`

`p[0] + p[1]                     # indexable like a plain tuple`

`33`

`x, y = p                        # unpack like a regular tuple`

`x, y`

`(11, 22)`

`p.x + p.y                       # fields also accessible by name`

`33`

`d = p._asdict()                 # convert to a dictionary`

`d['x']`

`11`

`Point(**d)                      # convert from a dictionary`

`(x=11, y=22)

` p._replace(x=100)`              # _replace() is like str.replace() but targets named fields`

`(x=100, y=22)`
`c:\users\ankit.shukla\appdata\local\programs\python\python39\lib\collections\__init__.py`


In [23]:
import collections

point = collections.namedtuple("point", ["x", "y"])

In [24]:
point.__doc__

'point(x, y)'

In [25]:
p = point(11, y=22)             # instantiate with positional args or keywords

In [26]:
p[0] + p[1]                     # indexable like a plain tuple

33

In [27]:
x, y = p                        # unpack like a regular tuple

In [32]:
x, y

(11, 22)

22

In [33]:
p.x + p.y                       # fields also accessible by name

33

In [34]:
d = p._asdict()                 # convert to a dictionary

In [35]:
d["x"]

11

In [36]:
d["y"]

22

In [38]:
point(**d)                      # convert from a dictionary

point(x=11, y=22)

In [39]:
p._replace(x=100) # _replace() is like str.replace() but targets named fields`

point(x=100, y=22)

In [40]:
p.x + p.y

33

In [41]:
p

point(x=11, y=22)

#### OrderedDict()

`The OrderedDict() function is similar to a normal dictionary object in Python. However, it remembers the order of the keys in which they were first inserted.`

In [42]:
import collections

d1 = collections.OrderedDict()
d1['A'] = 65
d1['C'] = 67
d1['B'] = 66
d1['D'] = 68

for k,v in d1.items():
    print (k,v)

A 65
C 67
B 66
D 68


Upon traversing the dictionary, pairs will appear in the order of their insertion.

In [43]:
d1

OrderedDict([('A', 65), ('C', 67), ('B', 66), ('D', 68)])

### deque()


`A deque object support appends and pops from either ends of a list. It is more memory efficient than a normal list object. In a normal list object, the removal of any item causes all items to the right to be shifted towards left by one index. Hence, it is very slow.`

In [44]:
q=collections.deque([10,20,30,40])
q.appendleft(0)

In [45]:
q

deque([0, 10, 20, 30, 40])

In [46]:
q.append(50)

In [47]:
q

deque([0, 10, 20, 30, 40, 50])

In [48]:
q.pop()

50

In [49]:
q

deque([0, 10, 20, 30, 40])

In [50]:
q.popleft()

0

In [51]:
q

deque([10, 20, 30, 40])

###  Python - Random Module

`The random module is a built-in module to generate the pseudo-random variables. It can be used perform some action randomly such as to get a random number, selecting a random elements from a list, shuffle elements randomly, etc.`

#### Generate Random Floats
The `random.random()` method returns a random float number between 0.0 to 1.0. The function doesn't need any arguments.



In [52]:
import random

In [54]:
random.random()

0.12823490816081018

#### Generate Random Integers

The `random.randint()` method returns a random integer between the specified integers.

In [55]:
import random

random.randint(1,20)

15

In [58]:
random.randint(1,50)

21

In [61]:
random.randint(2,3)

3

In [62]:
random.randint(3,3)

3

#### Generate Random Numbers within Range

The `random.randrange()` method returns a randomly selected element from the range created by the start, stop and step arguments. The value of start is 0 by default. Similarly, the value of step is 1 by default.

In [63]:
import random

random.randrange(1,10)

5

In [64]:
random.randrange(1,10, 2)

1

In [65]:
random.randrange(0, 101, 10)

40

#### Select Random Elements

The `random.choice()` method returns a randomly selected element from a non-empty sequence. An empty sequence as argument raises an IndexError.

In [66]:
import random
random.choice('computer')

'm'

In [67]:
random.choice([12,23,45,67,65,43])

45

In [68]:
random.choice([12,23,10,12,43])

10

#### Shuffle Elements Randomly

The `random.shuffle()` method randomly reorders the elements in a list.

In [69]:
import random 

numbers=[12,23,45,67,65,43]

random.shuffle(numbers)

In [70]:
 numbers

[23, 12, 67, 45, 43, 65]

as we can see that list elements order has been changed by shuffle built - in module.

In [71]:
numbers


[23, 12, 67, 45, 43, 65]